core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java

```diff
@@ -60,8 +60,10 @@
     public static final ParseField TYPE_FIELD = new ParseField("type");
     public static final ParseField QUERY_FIELD = new ParseField("query");
 
-    /** The default name for the match query */
-    public static final String NAME = "match";
+    /** The names for the match query. */
+    // TODO replace this with a Types like object
+    public static final Names NAMES = new Names("match", "match_phrase", "matchPhrase", "match_phrase_prefix", "matchPhrasePrefix",
```

Comment:
Personally I think the camelCase and snake_case handling should absolutely be done with ParseField as it then makes things really easy to do when we remove camelCase support since we only have to deal with changing ParseField. The Match Query seems to be the only query that has multiple names (ignoring snake and camel case variations) and actually we are abusing the name anyway by using it to switch between different features in the query. I think we should either:

1. Add a `type` parameter to the match query that has values `boolean` (default), `phrase` or `phrase_prefix` and deprecate all the query names except `match`
2. Split `MatchQueryBuilder` in to 3 classes which all use the same underlying class called `MatchQueryBuilder` (for the `boolean` type), `MatchPhraseQueryBuilder` (which extends MatchQueryBuilder but sets type to `phrase`), and `MatchPhrasePrefixQueryBuilder` (which extends MatchQueryBuilder but sets type to `phrase_prefix`).

Both of these approaches mean that all query builders will only have one name which makes sense since we really shouldn't have anything in Elasticsearch called by multiple names as its just confusing for both devs and users